src/history/hs_conn.c

```diff
@@ -113,6 +113,9 @@ __wt_hs_config(WT_SESSION_IMPL *session, const char **cfg)
     btree->file_max = (uint64_t)cval.val;
     WT_STAT_CONN_SET(session, cache_hs_ondisk_max, btree->file_max);
 
+    /* To access the history store file later, set up the connection with the HS open flag. */
+    F_SET(conn, WT_CONN_HS_OPEN);
```

Comment:
I think this deserves a bit of additional review, moving a flag set within its own function is unlikely to have much effect. @ajmorton  could you take a look?

Reply:
It's weird.. We're using release semantics* here where we don't set `WT_CONN_HS_OPEN` flag until the history store is properly started up, but there's no paired read barrier or acquires when we access `WT_CONN_HS_OPEN` in other threads. 

Prior to this PR the flag was only modified on connection open and connection close so the lack of pairing was largely harmless over the normal lifetime of a connection. The previous comment called out the stat server so perhaps that's the only thread that could race during HS open/close? Looking at our reads of `WT_CONN_HS_OPEN` I didn't see any accesses obviously linked with the stat server.

Making `WT_CONN_HS_OPEN` an atomic flag would resolve any issues since our atomics all use `__ATOMIC_SEQ_CST` and give us acquire/release semantics. It'll make the barrier unnecessary, but I wonder if keeping the barrier is useful as it communicates intentions more clearly than an atomic flag access? @luke-pearson I'll ping you on Slack to discuss.


*Technically this isn't release semantics as we're only ordering writes and not reads, this will be fixed as part of PM-3221's upcoming acquire/release upgrade and so isn't relevant to this PR.

Reply:
I realized the git diff rendering messed my review up a bit, this is getting moved into `__wt_hs_config`. Which is still called within `__wt_hs_open`. Anyway I thought about this for a bit and I'm pretty sure we don't need to worry about the implied race with the statistic server. 

This flag and almost every flag like it should be read eventually and as long as the other systems can cope with eventually we don't need to worry about ordering.

Realistically if anything depends on this flag being set it should be protected via a lock.

Reply:
Andrew pointed out rightly that this does match a pretty standard barrier case of:
Do a bunch of writes, in this case setting up the history store
Write barrier or release
Set the HS flag

But in that case we're missing ordering in every place that reads the flag.

Reply:
@luke-pearson, does it mean we need something to make sure we are reading the latest value of that flag and not a cached one?

Reply:
Best not to think about the cache, or latest values for that matter (this is actually a rather hard concept to grasp so don't feel bad about worrying about latest values) it might be easier to explain on a call. The situation Andrew and I are worried about hypothetically is:
```
// In this case its b-tree id is initially 0
history_store_table_btree_id = 10
write_barrier
F_SET(conn, WT_CONN_HS_OPEN)
```
The reader should read with a read barrier as follows:
```
hs_open = F_ISSET(conn, WT_CONN_HS_OPEN);
read_barrier
if (hs_open)
    read the b-tree id
```

Reply:
Hari did have this to say on the barrier:
> The barrier code is no longer needed. It was required before the history store was implemented. The function __wt_las_stats_update function updates the statistics based on the LAS_OPEN flag, which gets called from the stat log server. Due to this reason, the barrier was required. Now, with durable history, the stats are updated in a different location, and it also doesn't have any interaction with this flag.

But I think that disregards the other variables that could be potentially involved. I would like to think that accessing a table for the first time would require some kind of synchronization mechanism that would prevent the situation i described in the comment above from occurring.

Reply:
@luke-pearson and @ajmorton I didn't clearly understand the need for the write barrier here.

The code flows where this flag is set or unset happens in a single thread (recovery, shutdown), and there are no parallel activities that can occur while modifying this flag that need a barrier. Please let me know if I missed anything that needs a barrier to protect it.

Reply:
@kommiharibabu this is mostly about caution. One of the requirements we're moving towards in PM-3221 is any inter-thread lockless communication must have paired barriers since instruction reordering and visibility can get really hairy to reason about. It looks like this barrier is no longer needed, but we've had a lot of difficulty removing other barriers since we need to consider the worst case scheduling/instruction ordering scenarios.

The flag is set in a single thread, but we have a handful of other threads that read the flag to determine if they can perform HS operations. If we can guarantee these threads won't read `WT_CONN_HS_OPEN` even in the worst case then it's ok to remove the barrier. 
The easiest way to confirm this would be if all other threads are spawned by this thread after `WT_CONN_HS_OPEN` is set, and conversely that all other threads are joined before `WT_CONN_HS_OPEN` is set to false. In the past the stat server didn't follow this process so I'm not sure if other `WT_CONN_HS_OPEN` reading threads could be open and running already.

Reply:
If this code path is singly threaded I agree with @kommiharibabu we don't need a barrier.

Reply:
SGTM. Happy to drop the barrier in that case.
Thanks for double checking!